pkg/resource/deploy/source_eval.go

```diff
@@ -1248,10 +1387,129 @@ func (rm *resmon) RegisterResource(ctx context.Context,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	// We need to build the full alias spec list here, so we can pass it to transforms.
+	aliases := []*pulumirpc.Alias{}
+	for _, aliasURN := range req.GetAliasURNs() {
+		aliases = append(aliases, &pulumirpc.Alias{Alias: &pulumirpc.Alias_Urn{Urn: aliasURN}})
+	}
+
+	// We assume aliases are properly specified. However, if a request hasn't explicitly
+	// indicated that it is using properly specified aliases and the request is coming
+	// from Node.js, transform the aliases from the incorrect Node.js values to properly
+	// specified values, to maintain backward compatibility for users of older Node.js
+	// SDKs that aren't sending properly specified aliases.
+	transformAliases := !req.GetAliasSpecs() && requestFromNodeJS(ctx)
+
+	for _, aliasObject := range req.GetAliases() {
+		if transformAliases {
+			aliasObject = transformAliasForNodeJSCompat(aliasObject)
+		}
+		aliases = append(aliases, aliasObject)
+	}
+
+	var deleteBeforeReplace *bool
+	// Technically DeleteBeforeReplaceDefined should be used to decided if DeleteBeforeReplace should be looked at or
+	// not. However the Go sdk doesn't set Defined so we have a fallback here of respecting this field if either Defined
+	// is set or DeleteBeforeReplace is true.
+	if req.GetDeleteBeforeReplaceDefined() || req.GetDeleteBeforeReplace() {
+		deleteBeforeReplace = &req.DeleteBeforeReplace
+	}
+
+	opts := &pulumirpc.TransformationResourceOptions{
+		DependsOn:               req.GetDependencies(),
+		Protect:                 req.GetProtect(),
+		IgnoreChanges:           req.GetIgnoreChanges(),
+		ReplaceOnChanges:        req.GetReplaceOnChanges(),
+		Version:                 req.GetVersion(),
+		Aliases:                 aliases,
+		Provider:                req.GetProvider(),
+		Providers:               req.GetProviders(),
+		CustomTimeouts:          req.GetCustomTimeouts(),
+		PluginDownloadUrl:       req.GetPluginDownloadURL(),
+		RetainOnDelete:          req.GetRetainOnDelete(),
+		DeletedWith:             req.GetDeletedWith(),
+		DeleteBeforeReplace:     deleteBeforeReplace,
+		AdditionalSecretOutputs: req.GetAdditionalSecretOutputs(),
+		PluginChecksums:         req.GetPluginChecksums(),
+	}
+
+	// Before we calculate anything else run the transformations. First run the transforms for this resource,
+	// then it's parents etc etc
+	transforms, err := slice.MapError(req.Transformations, rm.wrapTransformCallback)
+	if err != nil {
+		return nil, err
+	}
+	for _, transform := range transforms {
+		newProps, newOpts, err := transform(ctx, name, string(t), custom, parent, props, opts)
+		if err != nil {
+			return nil, err
+		}
+		props = newProps
+		opts = newOpts
+	}
+	// Lookup our parents transformations and run those
+	err = func() error {
+		// Function exists to scope the lock
+		rm.resourceTransformationsLock.Lock()
+		defer rm.resourceTransformationsLock.Unlock()
+		rm.resGoalsLock.Lock()
+		defer rm.resGoalsLock.Unlock()
```

Comment:
Would it make sense to have these as RW mutexes instead?  If I understand correctly this lock is only for reading from `rm.resourceTransformations`, not to make sure only one transform call is happening at a time?

Reply:
Yeh I think we could